[WebAuthn] Start migrating to using modern AuthenticationServices API

pascoej opened the pull request:
#### db9ae698e3a18e4abfd0b2c8a54753de4b7407c3
<pre>
[WebAuthn] Start migrating to using modern AuthenticationServices API
<a href="https://bugs.webkit.org/show_bug.cgi?id=265703">https://bugs.webkit.org/show_bug.cgi?id=265703</a>
<a href="https://rdar.apple.com/113571336">rdar://113571336</a>

Reviewed by Brent Fulgham.

These AuthenticationServicesCore endpoints are becoming deprecated now that
you can perform requests through AuthenticationServices. This patch is to
start migrating WebKit to using the modern AuthenticationServices API.

This patch implements modal requests. Conditional meditation will be handled
in a follow-up patch.

The legacy AuthenticationServicesCore may still be used depending on
the WebAuthenticationASEnabled flag.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/Platform/spi/Cocoa/AuthenticationServicesCoreSPI.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(getAllLocalAuthenticatorCredentialsImpl):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesForwardDeclarations.h: Added.
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesSoftLink.h: Added.
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesSoftLink.mm: Added.
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(-[_WKASDelegate initWithPage:completionHandler:]):
(-[_WKASDelegate presentationAnchorForAuthorizationController:]):
(-[_WKASDelegate authorizationController:didCompleteWithAuthorization:]):
(-[_WKASDelegate authorizationController:didCompleteWithError:]):
(WebKit::toASUserVerificationPreference):
(WebKit::toAttestationConveyancePreference):
(WebKit::toASResidentKeyPreference):
(WebKit::toASAuthorizationSecurityKeyPublicKeyCredentialDescriptorTransport):
(WebKit::WebAuthenticatorCoordinatorProxy::constructASController):
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForRegisteration):
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForAssertion):
(WebKit::WebAuthenticatorCoordinatorProxy::performRequest):
(WebKit::toExceptionCode):
(WebKit::configureRegistrationRequestContext):
(WebKit::WebAuthenticatorCoordinatorProxy::contextForRequest):
(WebKit::continueAfterRequest):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.cpp:
(WebKit::WebAuthenticatorCoordinatorProxy::handleRequest):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/271525@main">https://commits.webkit.org/271525@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c716f611a73a268e4d18d0d14e068179136f968

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28677 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7320 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30048 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31298 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9437 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4681 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28947 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6052 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/24657 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5262 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25659 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32635 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/24799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26251 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/26104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31654 "layout-tests running") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/28005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6989 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/35294 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6858 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7624 "Passed tests") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/5893 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -848,6 +848,12 @@
 
+#if PLATFORM(IOS) \
+    || PLATFORM(MAC) \
+    || PLATFORM(VISION)
```
What about `IOS_SIMULATOR` or `MACCATALYST`? I suspect these don't work, but wanted to confirm.

pascoej replied:
Yes, they don't work. Only MAC, IOS, VISION as that's all that's supported by HAVE_UNIFIED_ASC_AUTH_UI.

brentfulgham commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -848,6 +848,12 @@
+#if PLATFORM(IOS) \
+    || PLATFORM(MAC) \
+    || PLATFORM(VISION)
+#define HAVE_WEBAUTHN_MODERN 1
```
This doesn't seem to be used in your patch. Was this replaced with `HAVE_WEBAUTHN_AS_MODERN`?

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesForwardDeclarations.h:
```diff
@@ -0,0 +1,328 @@
+
+OBJC_CLASS NSString;
+typedef NSString *ASAuthorizationPublicKeyCredentialUserVerificationPreference;
+typedef NSString *ASAuthorizationCustomMethod;
```
These forward declarations should probably live in the PAL folder somewhere.

pascoej replied:
I agree. I've filed https://bugs.webkit.org/show_bug.cgi?id=265814 to move the softlinks and forward declarations for this, AuthenticationServicesCoreSoftLink, and NearFieldSoftLink.

brentfulgham reviewed (approved):
I think this looks fine overall, but I saw a couple of things I think you should address before landing.

pascoej committed:
[WebAuthn] Start migrating to using modern AuthenticationServices API
https://bugs.webkit.org/show_bug.cgi?id=265703
rdar://113571336

Reviewed by Brent Fulgham.

These AuthenticationServicesCore endpoints are becoming deprecated now that
you can perform requests through AuthenticationServices. This patch is to
start migrating WebKit to using the modern AuthenticationServices API.

This patch implements modal requests. Conditional meditation will be handled
in a follow-up patch.

The legacy AuthenticationServicesCore may still be used depending on
the WebAuthenticationASEnabled flag.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/Platform/spi/Cocoa/AuthenticationServicesCoreSPI.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(getAllLocalAuthenticatorCredentialsImpl):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesForwardDeclarations.h: Added.
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesSoftLink.h: Added.
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesSoftLink.mm: Added.
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(-[_WKASDelegate initWithPage:completionHandler:]):
(-[_WKASDelegate presentationAnchorForAuthorizationController:]):
(-[_WKASDelegate authorizationController:didCompleteWithAuthorization:]):
(-[_WKASDelegate authorizationController:didCompleteWithError:]):
(WebKit::toASUserVerificationPreference):
(WebKit::toAttestationConveyancePreference):
(WebKit::toASResidentKeyPreference):
(WebKit::toASAuthorizationSecurityKeyPublicKeyCredentialDescriptorTransport):
(WebKit::WebAuthenticatorCoordinatorProxy::constructASController):
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForRegisteration):
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForAssertion):
(WebKit::WebAuthenticatorCoordinatorProxy::performRequest):
(WebKit::toExceptionCode):
(WebKit::configureRegistrationRequestContext):
(WebKit::WebAuthenticatorCoordinatorProxy::contextForRequest):
(WebKit::continueAfterRequest):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.cpp:
(WebKit::WebAuthenticatorCoordinatorProxy::handleRequest):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/271525@main
